### ISSUE TITLE
Using collapse component inside of a loop

### ISSUE THREAD

**Issue by ptums** [reactstrap/reactstrap#556]
Hello,

I am wondering what the correct way is to map over data and wrap each element inside of a collapse component. 

For example, I am iterating over a list of blog posts and when they are clicked on I want to display additional data about the post and I figure using collapse is the best way to hide the data until called to be displayed.

But when I try this I click on one of the blog posts and all collapse components reveal themselves. I understand this has to do with the toggle function. I am using the same function to toggle all the components. So my question is there a good way to either bind the toggle event to the individual component or is there a way to uniquely identify each component and bind that identifier to the toggle function?

This is my code so far. Thank you in advance for any advice or direction.

```
    return Object.keys(cat).map(index=>{
      let date = cat[index].createdAt.toString();
      date = date.substring(0, date.indexOf('T'));
      return(
        <ListGroupItem key={index}>
          <p onClick={this.toggle}><strong>{cat[index].title}</strong> - {date}</p>
          <Collapse isOpen={this.state.collapse}>{cat[index].title}</Collapse>
        </ListGroupItem>
      )
    });
```

**Comment by TheSharpieOne** [comment]
The best way to do something like this IMO would be to create a component which will manage the state for an individual `ListGroupItem` and it's collapse. Assuming you allow for multiple `Collapse`s to be showing at the same time, you can do something like this:
ListGroupCollapse:
```
import React from 'react';
import { ListGroupItem, Collapse } from 'reactstrap';

class ListGroupCollapse extends React.Component {
  constructor(props) {
    super(props);
    
    this.toggle = this.toggle.bind(this);
    this.state = {collapse: false};
  }
  
  toggle() {
    this.setState({ collapse: !this.state.collapse });
  }
  
  render() {
    const cat = this.props.cat;
    let date = cat.createdAt.toISOString();
    date = date.substring(0, date.indexOf('T'));
    return (
      <ListGroupItem>
        <div>
          <p onClick={this.toggle}>
            <strong>{cat.title}</strong> - {date}
          </p>
          <Collapse isOpen={this.state.collapse}>{cat.title}</Collapse>
        </div>
      </ListGroupItem>
    );
  }
}

export default ListGroupCollapse
```
And use it in your map like so:
```
        Object.keys(cat).map((key, index) =>
          <ListGroupCollapse key={index} cat={cat[key]} />
        )
```
Demo: https://www.webpackbin.com/bins/-Ksfqpl_i2sF2Lmjcv_Z

If you need it where opening one closes the others, you can do something similar with a callback (see this demo: https://www.webpackbin.com/bins/-Ksftn1S-gZ1WdzKkVCA)

Basically, you want to avoid creating/bind function in the render. Also creating the small ListGroupCollapse component which is only concerned about a single Collapse is easier to understand, plus it doesn't re-render when the props do no change.

**Comment by ptums** [comment]
Wow! Thank you this worked liked a charm.

**Comment by valvic66** [comment]
I have the same problem, could you pls advice?

import React from 'react'
import './index.css'

import {Collapse} from 'reactstrap'


class SubmenuGroup extends React.Component {
  getDefaultCSSClasses = () => {
    const classes = [this.props.classes || 'submenu_group']

    return classes
  }

  renderSubmenu = (submenuitems) => {
    return submenuitems.map ((submenuitem, i) => {
      return (
        <div 
          key={i}
        >
          {submenuitem}
        </div>  
      )
    })
  }
  
  render () {
    const {submenuitems, isSubmenuOpen} = this.props 

    return (
        <div className={this.getDefaultCSSClasses()}>
          {this.renderSubmenu(submenuitems)}
        </div>
    )
  }
}


export default class MenuGroup extends React.Component {
  constructor (props) {
    super (props)
    this.state = {
      isSubmenuOpen: false
    }
  }

  getDefaultCSSClasses = () => {
    const classes = [this.props.classes || 'menu_group']

    return classes
  }

  getDefaultCSSLogoClasses = () => {
    const { isSubmenuOpen } = this.state
    var classes = ''

    if (isSubmenuOpen) {
      classes = 'icon-chevron-down'
    } else {
      classes = 'icon-chevron-right'
    }

    return classes
  }

  handleLogoClick = (menu_row, i) => {
    console.log(menu_row, i)
    
    this.setState ({
      isSubmenuOpen: !this.state.isSubmenuOpen
    })
  }

  renderMenu = (menu_data) => {
    return menu_data.map ((menu_row, i) => {
      return (
        <div 
          key={i}
          className={this.getDefaultCSSClasses()}
        >
          <div className='menu_row'>
            <div>{menu_row.menuitem_name}</div>
            <div
              className={this.getDefaultCSSLogoClasses()}
              onClick={() => this.handleLogoClick(menu_row, i)}
            />
          </div>
          <Collapse isOpen={this.state.isSubmenuOpen}>
            <SubmenuGroup
              submenuitems={menu_row.submenuitems}
            />
          </Collapse>  
        </div>    
      )
    })
  }
  
  render () {
    const {menu_data} = this.props

    return (
      <div>
        {this.renderMenu(menu_data)}  
      </div>  
    )
  }
}

**Comment by valvic66** [comment]
```
import React from 'react'
import './index.css'

import {Collapse} from 'reactstrap'


class SubmenuGroup extends React.Component {
  getDefaultCSSClasses = () => {
    const classes = [this.props.classes || 'submenu_group']

    return classes
  }

  renderSubmenu = (submenuitems) => {
    return submenuitems.map ((submenuitem, i) => {
      return (
        <div 
          key={i}
        >
          {submenuitem}
        </div>  
      )
    })
  }
  
  render () {
    const {submenuitems, isSubmenuOpen} = this.props 

    return (
        <div className={this.getDefaultCSSClasses()}>
          {this.renderSubmenu(submenuitems)}
        </div>
    )
  }
}


export default class MenuGroup extends React.Component {
  constructor (props) {
    super (props)
    this.state = {
      isSubmenuOpen: false
    }
  }

  getDefaultCSSClasses = () => {
    const classes = [this.props.classes || 'menu_group']

    return classes
  }

  getDefaultCSSLogoClasses = () => {
    const { isSubmenuOpen } = this.state
    var classes = ''

    if (isSubmenuOpen) {
      classes = 'icon-chevron-down'
    } else {
      classes = 'icon-chevron-right'
    }

    return classes
  }

  handleLogoClick = (menu_row, i) => {
    console.log(menu_row, i)
    
    this.setState ({
      isSubmenuOpen: !this.state.isSubmenuOpen
    })
  }

  renderMenu = (menu_data) => {
    return menu_data.map ((menu_row, i) => {
      return (
        <div 
          key={i}
          className={this.getDefaultCSSClasses()}
        >
          <div className='menu_row'>
            <div>{menu_row.menuitem_name}</div>
            <div
              className={this.getDefaultCSSLogoClasses()}
              onClick={() => this.handleLogoClick(menu_row, i)}
            />
          </div>
          <Collapse isOpen={this.state.isSubmenuOpen}>
            <SubmenuGroup
              submenuitems={menu_row.submenuitems}
            />
          </Collapse>  
        </div>    
      )
    })
  }
  
  render () {
    const {menu_data} = this.props

    return (
      <div>
        {this.renderMenu(menu_data)}  
      </div>  
    )
  }
}
```

**Comment by valvic66** [comment]
I have the same problem, when I click the logo, all the submenus collapse. How to make each one collapse?

```
import React from 'react'
import './index.css'

import {Collapse} from 'reactstrap'


class SubmenuGroup extends React.Component {
  getDefaultCSSClasses = () => {
    const classes = [this.props.classes || 'submenu_group']

    return classes
  }

  renderSubmenu = (submenuitems) => {
    return submenuitems.map ((submenuitem, i) => {
      return (
        <div 
          key={i}
        >
          {submenuitem}
        </div>  
      )
    })
  }
  
  render () {
    const {submenuitems, isSubmenuOpen} = this.props 

    return (
        <div className={this.getDefaultCSSClasses()}>
          {this.renderSubmenu(submenuitems)}
        </div>
    )
  }
}


export default class MenuGroup extends React.Component {
  constructor (props) {
    super (props)
    this.state = {
      isSubmenuOpen: false
    }
  }

  getDefaultCSSClasses = () => {
    const classes = [this.props.classes || 'menu_group']

    return classes
  }

  getDefaultCSSLogoClasses = () => {
    const { isSubmenuOpen } = this.state
    var classes = ''

    if (isSubmenuOpen) {
      classes = 'icon-chevron-down'
    } else {
      classes = 'icon-chevron-right'
    }

    return classes
  }

  handleLogoClick = (menu_row, i) => {
    console.log(menu_row, i)
    
    this.setState ({
      isSubmenuOpen: !this.state.isSubmenuOpen
    })
  }

  renderMenu = (menu_data) => {
    return menu_data.map ((menu_row, i) => {
      return (
        <div 
          key={i}
          className={this.getDefaultCSSClasses()}
        >
          <div className='menu_row'>
            <div>{menu_row.menuitem_name}</div>
            <div
              className={this.getDefaultCSSLogoClasses()}
              onClick={() => this.handleLogoClick(menu_row, i)}
            />
          </div>
          <Collapse isOpen={this.state.isSubmenuOpen}>
            <SubmenuGroup
              submenuitems={menu_row.submenuitems}
            />
          </Collapse>  
        </div>    
      )
    })
  }
  
  render () {
    const {menu_data} = this.props

    return (
      <div>
        {this.renderMenu(menu_data)}  
      </div>  
    )
  }
}
```

**Comment by valvic66** [comment]
Could you pls advice on the above?

**Comment by valvic66** [comment]
Data comes via props:
```
    const menu_data = [
      {id: 0, menuitem_name: 'Home', submenuitems: ['Home1']},
      {id: 1, menuitem_name: 'Software', submenuitems: ['Software1', 'Software2']}
    ]
```

**Comment by ganesh-sankey** [comment]
@TheSharpieOne Thank you for code. The link that you have provided for DEMO no longer exists can you upload it again somewhere CodePen or other.

**Comment by TheSharpieOne** [comment]
I updated the links in the post above, also here they are:
https://stackblitz.com/edit/reactstrap-v5-4rezrr?file=Example.js
https://stackblitz.com/edit/reactstrap-v5-8f5phu?file=Example.js

**Comment by valvic66** [comment]
thanks a lot.

**Comment by Nasaakaa** [comment]
How to implement it to li tag?

**Comment by MAPESO** [comment]
Hi all 👋 I was facing the same problem a few hours ago. I would like to share the solution.

### Code that shows all the elements 👎 

```js
import React, { useState  } from 'react';
import { Collapse } from 'x.x.x';
const [isCollapse, setCollapse] = useState(false)

const toggle = () => setCollapse(!isCollapse);

const users = [ { name: 'Leo' }, { name: 'Mario'}, { name: 'Luc' }  ]

const listFriend = () => {
  return (
     <div>
        {users.map(item => {
          return (
            <button onClick={toggle}>Show more<button>
            <Collapse isOpen={isCollapse}>
                <h1>Hi !</h1>
           </Collapse>
          )
       })}
    </div>
   )
}
```

### Why does this happen 🤔 

The answer is quite simple, **we are not filtering item to be displayed**. In the method `toggle` we are not filtering absolutely nothing. If we want to show an item individually it will be necessary to filter it. With the following code the problem should be solved

### Code that will show an item individually 🎉 

```js

const users = [ { id: 1, name: 'Leo' }, { id: 2, name: 'Mario'}, { id: 3, name: 'Luc' }  ];

const toggle = index => {
    if (index) {
       // to show the rest of the elements
       setCollapse(!isCollapse);
    }
    if (index === 0) {
        //  to show first item
        setCollapse(!isCollapse);
      }
}

   <div>
        {users.map(item => {
          return (
            <button onClick={toggle(item.id)}>Show more<button>
            <Collapse isOpen={isCollapse}>
                <h1>Hi !</h1>
           </Collapse>
          )
       })}
    </div>
```
**NOTE:** This solution can be applied in any library that has the `Collapse` component : )

**Comment by vivipolli** [comment]
> Hi all wave I was facing the same problem a few hours ago. I would like to share the solution.
> ### Code that shows all the elements -1
> 
> ```js
> import React, { useState  } from 'react';
> import { Collapse } from 'x.x.x';
> const [isCollapse, setCollapse] = useState(false)
> 
> const toggle = () => setCollapse(!isCollapse);
> 
> const users = [ { name: 'Leo' }, { name: 'Mario'}, { name: 'Luc' }  ]
> 
> const listFriend = () => {
>   return (
>      <div>
>         {users.map(item => {
>           return (
>             <button onClick={toggle}>Show more<button>
>             <Collapse isOpen={isCollapse}>
>                 <h1>Hi !</h1>
>            </Collapse>
>           )
>        })}
>     </div>
>    )
> }
> ```
> 
> ### Why does this happen thinking
> 
> The answer is quite simple, **we are not filtering item to be displayed**. In the method `toggle` we are not filtering absolutely nothing. If we want to show an item individually it will be necessary to filter it. With the following code the problem should be solved
> ### Code that will show an item individually tada
> 
> ```js
> const users = [ { id: 1, name: 'Leo' }, { id: 2, name: 'Mario'}, { id: 3, name: 'Luc' }  ];
> 
> const toggle = index => {
>     if (index) {
>        // to show the rest of the elements
>        setCollapse(!isCollapse);
>     }
>     if (index === 0) {
>         //  to show first item
>         setCollapse(!isCollapse);
>       }
> }
> 
>    <div>
>         {users.map(item => {
>           return (
>             <button onClick={toggle(item.id)}>Show more<button>
>             <Collapse isOpen={isCollapse}>
>                 <h1>Hi !</h1>
>            </Collapse>
>           )
>        })}
>     </div>
> ```
> 
> **NOTE:** This solution can be applied in any library that has the `Collapse` component : )

it works without a library?

**Comment by imriven** [comment]
Is there a way to use both collapse and uncontrolled collapse?  I'm mapping over some data and using collapse and it works fine until you get down to mobile size. Then I have that whole element disappear for the mobile element. The mobile is using uncontrolled collapse because in the mobile state I need multiple collapses. 

I have multiple collapses to use until the browser hits a certain breakpoint. So I'm using uncontrolled collapse for that so I can use Ids. Well from commenting various things out, the collapse even though it's hidden and the elements that use it are hidden, it's messing with my styling. The uncontrolled collapses still work but none of my styling works at all. It's even affecting styling on other pages which I'm not even sure how that's possible. Is there some work around I've been at it for hours and I'm out of ideas....
